### PR TITLE
Papilo primal/dual crush

### LINKED ISSUE #513: [FEA] Restore initial user-provided solution support w/ papilo presolve
Currently, our facilities for reading initial user-provided solutions from a folder are no longer supported.
This is caused by the lack of support by papilo for converting solutions from the original problem space to the reduced problem space.
[As mentioned by this issue in the Papilo repo](https://github.com/scipopt/papilo/issues/36), this is a non-trivial task that would require disabling some presolvers that cut off some solutions.

Papilo reductions are stored in `PostsolvedStorage` as an ordered sequence of [specific reduction types](https://github.com/scipopt/papilo/blob/main/src/papilo/core/postsolve/ReductionType.hpp). It should be possible to take inspiration from the `undo()` method of Postsolve to implement the inverse equivalent of these transformations to convert a solution from the original problem space to the reduced one. 

Additionally, it should be possible to take inspiration from SCIP's code which uses Papilo as well.

### LINKED ISSUE #1060: [FEA] Convert solution in original space to solution in presolve space with Papilo
This is currently not supported by Papilo. We could push this improvement upstream.

### PR DESCRIPTION
**Pull request by aliceb-nv**
This PR implements support for crushing primal incumbents in MIP mode into the Papilo problem space, and crushing primal/dual vectors for LP.

A bugfix is also included to allow consecutive solves to be run in the same GTest process without corrupting the OpenMP runtime.

Closes https://github.com/NVIDIA/cuopt/issues/513
Closes https://github.com/NVIDIA/cuopt/issues/1060

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/cuopt/blob/HEAD/CONTRIBUTING.md).
- Testing
   - [ ] New or existing tests cover these changes
   - [ ] Added tests
   - [ ] Created an issue to follow-up
   - [ ] NA
- Documentation
   - [ ] The documentation is up to date with these changes
   - [ ] Added new documentation
   - [ ] NA

### CODE REVIEW

File: cpp/src/mip_heuristics/presolve/third_party_presolve.cpp [hunk 1]
```diff
@@ -876,6 +881,242 @@ void third_party_presolve_t<i_t, f_t>::uncrush_primal_solution(
...
+ * This is the forward counterpart of Papilo's Postsolve::undo(). It replays
+ * each presolve reduction in forward order to transform variable/dual values,
+ * then projects onto the surviving columns/rows via origcol/origrow_mapping.
+ *
+ * Only two reductions actually transform survivor coordinates:
+ *   kParallelCol             — merges x[col1] into x[col2]
+ *   kRowBoundChangeForcedByRow — conditionally transfers y[deleted_row] → y[kept_row]
+ */
+template <typename i_t, typename f_t>
+void third_party_presolve_t<i_t, f_t>::crush_primal_dual_solution(
```
Comment (rg20) [inline cpp/src/mip_heuristics/presolve/third_party_presolve.cpp:922]: Can we push this to Papilo itself?  

When new presolvers (or reductions) are added, this can get outdated.

Comment (aliceb-nv) [reply cpp/src/mip_heuristics/presolve/third_party_presolve.cpp:922]: That's on my radar :) There are a couple of PRs I will be pushing to Papilo

File: cpp/src/mip_heuristics/diversity/diversity_manager.cu [hunk 2]
```diff
@@ -418,6 +436,9 @@ solution_t<i_t, f_t> diversity_manager_t<i_t, f_t>::run_solver()
     CUOPT_LOG_INFO("GPU heuristics disabled via CUOPT_DISABLE_GPU_HEURISTICS=1");
     population.initialize_population();
     population.allocate_solutions();
+    if (check_b_b_preemption()) { return population.best_feasible(); }
+    add_user_given_solutions(initial_sol_vector);
+    population.add_solutions_from_vec(std::move(initial_sol_vector));
```
[the remark this answers is not in the document]
Comment (aliceb-nv) [reply cpp/src/mip_heuristics/diversity/diversity_manager.cu]: Minor edge-case in a debugging-only path. We can safely ignore.

**Comment by nguidotti** [comment]
Moving to 26.08

File: cpp/src/mip_heuristics/presolve/third_party_presolve.cpp [hunk 3]
```diff
@@ -574,6 +574,11 @@ void set_presolve_methods(papilo::Presolve<f_t>& presolver,
   presolver.addPresolveMethod(uptr(new papilo::Probing<f_t>()));
 
   if (!dual_postsolve) {
+    // SingletonStuffing causes dual crushing failures on:
```
Comment (rg20) [inline cpp/src/mip_heuristics/presolve/third_party_presolve.cpp:580]: There is no dual crush for MIP right?

File: cpp/src/mip_heuristics/presolve/third_party_presolve.cpp [hunk 4]
```diff
@@ -876,6 +881,250 @@ void third_party_presolve_t<i_t, f_t>::uncrush_primal_solution(
...
+  const bool crush_rc = !z_original.empty() && crush_dual;
+  if (crush_rc) { cuopt_assert((int)z_original.size() == (int)storage.nColsOriginal, ""); }
+
+  std::vector<f_t> x(x_original.begin(), x_original.end());
+  std::vector<f_t> y(y_original.begin(), y_original.end());
+  std::vector<f_t> z(z_original.begin(), z_original.end());
+
+  // Track current coefficient values for entries modified by kCoefficientChange,
+  // so repeated changes to the same (row, col) are handled correctly.
+  std::map<std::pair<int, int>, f_t> coeff_current;
```
Comment (rg20) [inline cpp/src/mip_heuristics/presolve/third_party_presolve.cpp]: This can be very slow.  

Can you avoid map?  Can you just store a CSR/COO of the size of the original matrix?

File: cpp/src/mip_heuristics/solve.cu [hunk 5]
```diff
@@ -334,6 +338,10 @@ template <typename i_t, typename f_t>
 mip_solution_t<i_t, f_t> solve_mip_helper(optimization_problem_t<i_t, f_t>& op_problem,
                                           mip_solver_settings_t<i_t, f_t> const& settings_const)
 {
+  // Release the OMP thread pool on exit so that consecutive solves from different
+  // std::async threads (e.g. in tests) don't deadlock on stale OMP master affinity.
+  auto omp_guard = cuopt::scope_guard([] { omp_pause_resource_all(omp_pause_hard); });
```
Comment (nguidotti) [inline cpp/src/mip_heuristics/solve.cu]: This fix can be removed since we migrated everything to OpenMP already.

File: cpp/src/mip_heuristics/presolve/third_party_presolve.cpp (hunk 4 above)
Comment (aliceb-nv) [reply cpp/src/mip_heuristics/presolve/third_party_presolve.cpp]: The number of kCoefficientChange entries is usually << nnz, so I defaulted to using a sparse COO but a std::map is indeed not great
switched to a unordered_map w/ proper hash

File: cpp/tests/linear_programming/unit_tests/presolve_test.cu [hunk 6]
```diff
@@ -383,6 +491,416 @@ TEST(pslp_presolve, postsolve_multiple_problems)
...
+  double obj_tol = 1e-3 * (1.0 + std::abs(cold_obj));
+  EXPECT_NEAR(warm_obj, cold_obj, obj_tol) << "warmstarted objective should match cold solve";
+
+  EXPECT_LT(warm_iters, cold_iters)
+    << "warmstarted solve should not take more iterations than cold solve"
+    << " (cold=" << cold_iters << ", warm=" << warm_iters << ")";
+}
+
+// clang-format off
+INSTANTIATE_TEST_SUITE_P(
```
Comment (nguidotti) [inline cpp/tests/linear_programming/unit_tests/presolve_test.cu:864]: NITPICK: do we need to test that many instances?

**Review by nguidotti — APPROVED** [review]
Looks good to me! Thanks for the hard work, Alice!

File: cpp/tests/linear_programming/unit_tests/presolve_test.cu (hunk 6 above)
Comment (aliceb-nv) [reply cpp/tests/linear_programming/unit_tests/presolve_test.cu:864]: Test runs are fast and crushing is non-trivial and has plenty of edge cases, so the rationale here is better safe than sorry :)

File: datasets/mip/download_miplib_test_dataset.sh [hunk 7]
```diff
@@ -26,6 +26,37 @@ INSTANCES=(
     "enlight11"
     "supportcase22"
     "pk1"
+    "app1-1"
```
Comment (ramakrishnap-nv) [inline datasets/mip/download_miplib_test_dataset.sh:29]: Are we testing for all these datasets ?

Comment (aliceb-nv) [reply datasets/mip/download_miplib_test_dataset.sh:29]: Yes
These are smaller instances, and they're used for a quick running test (the crushing roundtrip). This is important because edge cases and numerical instability mean that ensuring the correctness of the crushing function is difficult without diverse test instances

**Review by ramakrishnap-nv — APPROVED** [review]
